### PR TITLE
TW-1760: Create `UploadManager` to manage upload file

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket 

- #1760 

### Need merge:

- https://github.com/linagora/matrix-dart-sdk/pull/64

#### `UploadManager`
- [x] `UploadManager` to manage upload file
- [x] Send fake event first
- [x] Add task to queue 
- [x] Handle upload file success
- [x] Error handling
- [x] Write unit test

#### Improve UI/UX upload load files (Mobile/web)

- [x] Display cancel token and progress upload
- [x] Update correctly uploading file
- [x] Cancel uploading file => remove fake event
- [x] Error handling upload fail

#### Improve UI/UX upload media (Mobile/web)

- [x] Gen and upload thumbnail
- [x] Update correctly uploading video/image
- [x] Display cancel token and progress upload
- [x] Cancel uploading file => remove fake event
- [x] Error handling upload fail

#### Mobile

- Upload file

https://github.com/linagora/twake-on-matrix/assets/99852347/58e0569f-8f7b-4762-a4a8-f67db0f61c83

https://github.com/linagora/twake-on-matrix/assets/99852347/581f16b9-ce83-488e-8ead-ea98f9c72dc8

- Upload media

https://github.com/linagora/twake-on-matrix/assets/99852347/db74b1e5-f06e-46d7-b215-916f4f5dec2f

https://github.com/linagora/twake-on-matrix/assets/99852347/262217c4-0b1c-4f71-9567-1711c83a266a

### WEB

- Upload file

https://github.com/linagora/twake-on-matrix/assets/99852347/a68d5ffe-1834-494c-8602-1158c0afc3c2

https://github.com/linagora/twake-on-matrix/assets/99852347/c492f972-577e-4133-a361-c841d98f9226

https://github.com/linagora/twake-on-matrix/assets/99852347/9775dd81-ee6b-479e-be90-6d4abdd7bf74

### CODE REVIEW

File: lib/domain/usecase/send_file_interactor.dart [hunk 1]
```diff
@@ -18,13 +18,25 @@ class SendFileInteractor {
         final txid = room.storePlaceholderFileInMem(
           fileInfo: fileInfo,
         );
-        room.sendFileEvent(
+
+        final fakeImageEvent = await room.sendFakeImagePickerFileEvent(
```
Comment (sherlockvn) [inline lib/domain/usecase/send_file_interactor.dart:22]: `fakeImageEvent` -> `fakeFileEvent`

File: lib/domain/usecase/send_file_interactor.dart [hunk 2]
```diff
@@ -18,13 +18,25 @@ class SendFileInteractor {
...
+        );
+
+        room.sendFileEventMobile(
           fileInfo,
           msgType: fileInfo.msgType,
           txid: txid,
           editEventId: editEventId,
           inReplyTo: inReplyTo,
           shrinkImageMaxDimension: shrinkImageMaxDimension,
+          fakeImageEvent: fakeImageEvent,
```
Comment (sherlockvn) [inline lib/domain/usecase/send_file_interactor.dart:39]: fakeImageEvent -> fakeFileEvent

File: lib/domain/usecase/send_file_on_web_interactor.dart [hunk 3]
```diff
@@ -1,25 +0,0 @@
-import 'package:fluffychat/presentation/extensions/send_file_web_extension.dart';
```
Comment (sherlockvn) [inline lib/domain/usecase/send_file_on_web_interactor.dart:1]: delete the file

File: lib/domain/usecase/send_files_on_web_with_caption_interactor.dart [hunk 4]
```diff
@@ -1,63 +0,0 @@
-import 'package:fluffychat/presentation/extensions/send_file_extension.dart';
```
Comment (sherlockvn) [inline lib/domain/usecase/send_files_on_web_with_caption_interactor.dart:1]: delete it

File: lib/pages/chat/events/images_builder/sending_image_info_widget.dart [hunk 5]
```diff
@@ -71,9 +81,14 @@ class SendingImageInfoWidget extends StatelessWidget {
                   strokeWidth: 2,
                   color: LinagoraRefColors.material().primary[100],
                 ),
-                Icon(
-                  Icons.close,
-                  color: LinagoraRefColors.material().primary[100],
+                InkWell(
+                  child: Icon(
+                    Icons.close,
+                    color: LinagoraRefColors.material().primary[100],
```
Comment (sherlockvn) [inline lib/pages/chat/events/images_builder/sending_image_info_widget.dart:87]: move it to style file

File: lib/pages/chat/events/images_builder/sending_image_info_widget.dart [hunk 6]
```diff
@@ -28,19 +29,28 @@ class SendingImageInfoWidget extends StatelessWidget {
...
+  @override
+  State<SendingImageInfoWidget> createState() => _SendingImageInfoWidgetState();
+}
+
+class _SendingImageInfoWidgetState extends State<SendingImageInfoWidget>
+    with UploadFileMixin {
+  @override
+  Event get event => widget.event;
+
+  final ValueNotifier<double> sendingFileProgressNotifier = ValueNotifier(0);
```
Comment (sherlockvn) [inline lib/pages/chat/events/images_builder/sending_image_info_widget.dart:43]: dispose it

File: lib/domain/usecase/send_file_interactor.dart (hunk 1 above)
Comment (nqhhdev) [reply lib/domain/usecase/send_file_interactor.dart:22]: Not use remove `send_file_interactor.dart`

File: lib/domain/usecase/send_file_interactor.dart (hunk 2 above)
Comment (nqhhdev) [reply lib/domain/usecase/send_file_interactor.dart:39]: Not use remove `send_file_interactor.dart`

File: lib/utils/manager/upload_manager/models/upload_file_info.dart [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  UploadFileInfo({
+    required super.txid,
+    required this.createdAt,
+    required this.uploadStateStreamController,
+    required this.uploadStream,
+    required this.cancelToken,
+  });
+
+  @override
+  List<Object?> get props => [
```
Comment (sherlockvn) [inline lib/utils/manager/upload_manager/models/upload_file_info.dart:23]: add `createdAt` also

**Comment by hieutbui** [comment]
Tested on Windows and it works well

**Review by hieutbui — APPROVED** [review]

File: lib/pages/chat/events/images_builder/sending_image_info_widget.dart (hunk 6 above)
Comment (Te-Z) [reply lib/pages/chat/events/images_builder/sending_image_info_widget.dart:49]: @sherlockvn already disposed line 45

**Comment by Te-Z** [comment]
There's an error when trying to send a file on web. I think that's why the upload looks laggy at first, before displaying the loading animation

[Capture vidéo du 08-07-2024 17:29:51.webm](https://github.com/linagora/twake-on-matrix/assets/31937920/247e8ffe-d4ff-4c8b-bb5e-ae79462cf57a)

**Comment by nqhhdev** [comment]
> There's an error when trying to send a file on web. I think that's why the upload looks laggy at first, before displaying the loading animation
> 
>  Capture.video.du.08-07-2024.17.29.51.webm

This is an error for `Dialog`, not related to  `UploadManager`. If you found issue pls create a ticket for this

**Review by sherlockvn — APPROVED** [review]

**Comment by sherlockvn** [comment]
I tested in iOS platform and web platform and its worked well for me.

**Review by sherlockvn — APPROVED** [review]

**Comment by Te-Z** [comment]
> > There's an error when trying to send a file on web. I think that's why the upload looks laggy at first, before displaying the loading animation
> > Capture.video.du.08-07-2024.17.29.51.webm
> 
> This is an error for `Dialog`, not related to `UploadManager`. If you found issue pls create a ticket for this

Still the upload seems laggy @nqhhdev : lot of time before having the loading animation. Also this ticket is about upload and the error is about upload dialog. There's already 244 issues here.

**Comment by Te-Z** [comment]
also @nqhhdev there's no error on main branch but it happens on your branch. Please fix this.

**Comment by nqhhdev** [comment]
> > > There's an error when trying to send a file on web. I think that's why the upload looks laggy at first, before displaying the loading animation
> > > Capture.video.du.08-07-2024.17.29.51.webm
> > 
> > 
> > This is an error for `Dialog`, not related to `UploadManager`. If you found issue pls create a ticket for this
> 
> Still the upload seems laggy @nqhhdev : lot of time before having the loading animation. Also this ticket is about upload and the error is about upload dialog. There's already 244 issues here.

About laggy: We are fix it on pr https://github.com/linagora/twake-on-matrix/pull/1898, pls check it.

**Comment by nqhhdev** [comment]
> also @nqhhdev there's no error on main branch but it happens on your branch. Please fix this.

Currently, last commit of `feature/upload_file` like image below

Screenshot 2024-07-09 at 14 59 06

- On `main` branch, we fixed a memory leak for `file_picker`. `feature/upload_file` There are no latest updates for the `picker_file` section, so the two branches are different.

**Review by Te-Z — APPROVED** [review]

**Pull request merged**